### PR TITLE
add HTTP auth support Closes #159

### LINKED ISSUE #159: add HTTP auth support
We should probably show a simple auth dialog on:
WebViewClient.onReceivedHttpAuthRequest()
https://developer.android.com/reference/android/webkit/WebViewClient.html#onReceivedHttpAuthRequest(android.webkit.WebView,%20android.webkit.HttpAuthHandler,%20java.lang.String,%20java.lang.String)

We currently show an ugly Android error page when trying to load a page that requires authentication. Webkit then _doesn't_ call WebViewClient.onReceivedError(), so we'd still need to at least redirect to our own error page even if we don't add http auth support.

It looks like iOS doesn't support this yet:
https://bugzilla.mozilla.org/show_bug.cgi?id=1318896

### PR DESCRIPTION
**Pull request by kai-chi**
This PR adds support for websites with HTTP authentication. Method onReceivedHttpAuthRequest creates a dialog to input username and password. 
Attaching a screenshot of how the window looks like

The credentials are not explicitly stored by me but they are not erased as the rest of the session data. It is not cleared with WebViewDatabase.clearHttpAuthUsernamePassword(). I struggle to figure out where it is stored and how I can clear it out.

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 1]
```diff
@@ -30,21 +30,13 @@
 import android.support.v4.widget.SwipeRefreshLayout;
 import android.text.TextUtils;
 import android.util.Log;
-import android.view.Gravity;
-import android.view.LayoutInflater;
-import android.view.View;
-import android.view.ViewGroup;
-import android.view.Window;
-import android.view.WindowManager;
+import android.view.*;
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: \[Checkstyle\] ERROR: Using the '\.\*' form of import should be avoided \- android\.view\.\*\.

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 2]
```diff
@@ -30,21 +30,13 @@
...
 import android.webkit.CookieManager;
 import android.webkit.URLUtil;
-import android.widget.FrameLayout;
-import android.widget.ImageButton;
-import android.widget.ImageView;
-import android.widget.PopupWindow;
-import android.widget.TextView;
-import android.widget.Toast;
-
+import android.widget.*;
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: \[Checkstyle\] ERROR: Using the '\.\*' form of import should be avoided \- android\.widget\.\*\.

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 3]
```diff
@@ -58,18 +50,10 @@
 import org.mozilla.focus.observer.AverageLoadTimeObserver;
 import org.mozilla.focus.open.OpenWithFragment;
 import org.mozilla.focus.popup.PopupUtils;
-import org.mozilla.focus.session.NullSession;
-import org.mozilla.focus.session.Session;
-import org.mozilla.focus.session.SessionCallbackProxy;
-import org.mozilla.focus.session.SessionManager;
-import org.mozilla.focus.session.Source;
+import org.mozilla.focus.session.*;
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: \[Checkstyle\] ERROR: Using the '\.\*' form of import should be avoided \- org\.mozilla\.focus\.session\.\*\.

File: app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java [hunk 4]
```diff
@@ -58,18 +50,10 @@
...
 import org.mozilla.focus.session.ui.SessionsSheetFragment;
 import org.mozilla.focus.telemetry.TelemetryWrapper;
-import org.mozilla.focus.utils.Browsers;
-import org.mozilla.focus.utils.Features;
-import org.mozilla.focus.utils.StatusBarUtils;
-import org.mozilla.focus.utils.SupportUtils;
-import org.mozilla.focus.utils.UrlUtils;
+import org.mozilla.focus.utils.*;
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/fragment/BrowserFragment.java]: \[Checkstyle\] ERROR: Using the '\.\*' form of import should be avoided \- org\.mozilla\.focus\.utils\.\*\.

File: app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                .setPositiveButton(R.string.action_ok, new DialogInterface.OnClickListener() {
+                    public void onClick(DialogInterface dialog, int whichButton) {
+                        if (mOkListener != null) {
+                            mOkListener.onOk(mHost, mRealm, getUsername(), getPassword());
+                        }
+                    }
+                })
+                .setNegativeButton(R.string.action_cancel, new DialogInterface.OnClickListener() {
+                    public void onClick(DialogInterface dialog, int whichButton) {
+                        if (mCancelListener != null) mCancelListener.onCancel();
```
Comment (sputnikci) [inline app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java]: \[Checkstyle\] ERROR: 'if' construct must use '\{\}'s\.

File: app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                    }
+                })
+                .setOnCancelListener(new DialogInterface.OnCancelListener() {
+                    public void onCancel(DialogInterface dialog) {
+                        if (mCancelListener != null) mCancelListener.onCancel();
```
Comment (sputnikci) [inline app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java]: \[Checkstyle\] ERROR: 'if' construct must use '\{\}'s\.

File: app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+
+    public void show() {
+        mDialog.show();
+        mUsernameView.requestFocus();
+    }
+
+
+    private void createDialog() {
```
Comment (sputnikci) [inline app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java]: \[PMD\] INFO: The method 'createDialog\(\)' has a NCSS line count of 22\.

Comment (sputnikci) [inline app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java]: \[PMD\] INFO: The method 'createDialog\(\)' has a cyclomatic complexity of 5\.

File: app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java (hunk 5 above)
Comment (sputnikci) [inline app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java]: \[PMD\] INFO: Avoid using if statements without curly braces

File: app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java (hunk 6 above)
Comment (sputnikci) [inline app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java]: \[PMD\] INFO: Avoid using if statements without curly braces

File: app/src/webview/java/org/mozilla/focus/webview/TrackingProtectionWebViewClient.java [hunk 8]
```diff
@@ -9,17 +9,14 @@
 import android.net.Uri;
 import android.os.AsyncTask;
 import android.support.annotation.WorkerThread;
-import android.webkit.WebResourceRequest;
-import android.webkit.WebResourceResponse;
-import android.webkit.WebView;
-import android.webkit.WebViewClient;
-
+import android.webkit.*;
```
Comment (sputnikci) [inline app/src/webview/java/org/mozilla/focus/webview/TrackingProtectionWebViewClient.java]: \[Checkstyle\] ERROR: Using the '\.\*' form of import should be avoided \- android\.webkit\.\*\.

**Comment by pocmo** [comment]
Pinging @brampitoyo regarding UX.

File: app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    private OkListener mOkListener;
+    private CancelListener mCancelListener;
+
+
+    public HttpAuthenticationDialog(Context context, String host, String realm) {
+        mContext = context;
+        mHost = host;
+        mRealm = realm;
+        createDialog();
```
Comment (pocmo) [inline app/src/webview/java/org/mozilla/focus/webview/HttpAuthenticationDialog.java]: Creating the dialog from the constructor is not very clean. I'd expect an API more like:

Builder()
 .set..
 .set..
 .build()

Almost like AlertDialog.Builder but for this purpose here.

File: app/src/webview/java/org/mozilla/focus/webview/SystemWebView.java [hunk 10]
```diff
@@ -55,7 +55,7 @@
     public SystemWebView(Context context, AttributeSet attrs) {
         super(context, attrs);
 
-        client = new FocusWebViewClient(getContext().getApplicationContext());
+        client = new FocusWebViewClient(context);
```
Comment (pocmo) [inline app/src/webview/java/org/mozilla/focus/webview/SystemWebView.java]: This is a risky change. Are you sure nothing is keeping a long-lived reference to this context?

File: app/src/webview/java/org/mozilla/focus/webview/TrackingProtectionWebViewClient.java [hunk 11]
```diff
@@ -106,7 +109,7 @@ public WebResourceResponse shouldInterceptRequest(final WebView view, final WebR
         if ((!request.isForMainFrame()) &&
                 currentPageURL != null &&
                 matcher.matches(resourceUri, Uri.parse(currentPageURL))) {
-                // Bandaid for issue #26: currentPageUrl can still be null, and needs to be investigated further.
+            // Bandaid for issue #26: currentPageUrl can still be null, and needs to be investigated further.
```
Comment (pocmo) [inline app/src/webview/java/org/mozilla/focus/webview/TrackingProtectionWebViewClient.java]: nit: unrelated change..

File: app/src/webview/java/org/mozilla/focus/webview/TrackingProtectionWebViewClient.java [hunk 12]
```diff
@@ -137,4 +140,28 @@ public void onPageStarted(WebView view, String url, Bitmap favicon) {
...
+        httpAuthenticationDialog.setCancelListener(new HttpAuthenticationDialog.CancelListener() {
+            @Override
+            public void onCancel() {
+                handler.cancel();
+                httpAuthenticationDialog = null;
+            }
+        });
+
+        httpAuthenticationDialog.show();
+    }
```
Comment (pocmo) [inline app/src/webview/java/org/mozilla/focus/webview/TrackingProtectionWebViewClient.java]: This is the wrong place for this. IWebView.Callback should provide a generic on** method for that and this is the method we should call here. Background is: There are multiple implementations of IWebView and we do. not want to add http-auth-dialogs to all of them. So instead we pass this along. With that we only need to implement the http-auth-dialog in one place for all implementations of IWebView. Luckily the callback will be executed in a fragment and with that you do not need all the passing around of the Context here.

**Comment by brampitoyo** [comment]
This is what the UI should look like, with measurements:



* Spacing and dimension should be identical to the Settings menu and page modal dialogues.
* Highlight and button colour values should be `blue-50`: `#0a84ff`
* Heading font should be identical to the size of our menu items: `16dp`. Its weight should also be `semibold`.

**Comment by kai-chi** [comment]
thanks @sputnikci and @pocmo for your comments and @brampitoyo for the layout. attaching screenshot of the current status.

File: app/src/main/java/org/mozilla/focus/web/HttpAuthenticationDialogBuilder.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    public void show() {
+        dialog.show();
+        int photonBlue = ContextCompat.getColor(context, R.color.photonBlue50);
+        dialog.getButton(DialogInterface.BUTTON_POSITIVE).setTextColor(photonBlue);
+        dialog.getButton(DialogInterface.BUTTON_NEGATIVE).setTextColor(photonBlue);
+        usernameTextView.requestFocus();
+    }
+
+
+    public void createDialog() {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/web/HttpAuthenticationDialogBuilder.java]: \[PMD\] INFO: The method 'createDialog\(\)' has a NCSS line count of 12\.

File: app/src/main/java/org/mozilla/focus/web/HttpAuthenticationDialogBuilder.java [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+                    return true;
+                }
+                return false;
+            }
+        });
+
+        buildDialog(view);
+    }
+
+    private void buildDialog(View view) {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/web/HttpAuthenticationDialogBuilder.java]: \[PMD\] INFO: The method 'buildDialog\(View\)' has a NCSS line count of 12\.

File: app/src/main/java/org/mozilla/focus/web/IWebView.java [hunk 15]
```diff
@@ -75,6 +76,8 @@ public HitTarget(final boolean isLink, final String linkURL, final boolean isIma
         void resetBlockedTrackers();
 
         void onBlockingStateChanged(boolean isBlockingEnabled);
+
+        void onHttpAuthRequest(HttpAuthHandler handler, String host, String realm);
```
Comment (pocmo) [inline app/src/main/java/org/mozilla/focus/web/IWebView.java]: HttpAuthHandler is a specific implementation for WebView. We'd need to hide that behind an interface that IWebView defines.

**Comment by kai-chi** [comment]
thanks, @pocmo for the comment. I applied the changes. 
One thing. sputnik has requested two changes that I find hard to address. I don't see the benefit of splitting those methods. Is it a must to address those comments?

File: app/src/main/java/org/mozilla/focus/web/IWebView.java [hunk 16]
```diff
@@ -9,6 +9,7 @@
 import android.support.annotation.Nullable;
 import android.view.View;
 
+import android.webkit.HttpAuthHandler;
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/web/IWebView.java]: \[Checkstyle\] ERROR: Unused import \- android\.webkit\.HttpAuthHandler\.

**Comment by kai-chi** [comment]
hey, should I complete something to get this PR reviewed?

**Comment by kai-chi** [comment]
@pocmo anything more to add?

**Comment by kai-chi** [comment]
can we change the label from 'needs review' to probably 'needs merge' since it has been reviewed? :)

**Comment by boek** [comment]
Squashed all the commits into: #2646

**Comment by bbinto** [comment]
@boek+ @Sdaswani - is this something we even want to support, isn't that less secure and we potentially support insecure authentication (since it's not HTTPs)?

**Comment by Sdaswani** [comment]
@bbinto I think this is the standard part of the web and adding support doesn't make us less secure - the user is the ultimate driver of the decision to use this auth method. It expands our coverage of the web. I say we merge, especially since @pocmo has reviewed the code over several passes.

**Comment by bbinto** [comment]
I had hopes you'd say that and confirm. Thanks, let's do it!

**Comment by boek** [comment]
@bbinto Also this is for authentication, which can be over https

**Comment by npark-mozilla** [comment]
Tested it on http://www.httpwatch.com
